<a href="https://colab.research.google.com/github/poojajangid19/Data-Analyst/blob/main/IMDB_case_study.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## IMDB Movie Business Use-case

#### Imagine that you are working as a Data Scientist for an Analytics firm

- And you are analysing some movie trends for a client
- IMDB has online database of information related to movies
- The database contains info of several years about:
    - Movies
    - Rating
    - Director
    - Popularity
    - Revenue & Budget



In [ ]:
import pandas as pd
import numpy as np


#### Lets download and read the IMDB dataset

  - Lets first download the dataset
  - File1: https://drive.google.com/file/d/1s2TkjSpzNc4SyxqRrQleZyDIHlc7bxnd/view?usp=sharing
  - File2: https://drive.google.com/file/d/1Ws-_s1fHZ9nHfGLVUQurbHDvStePlEJm/view?usp=sharing

In [ ]:
!pip install gdown
!gdown 1s2TkjSpzNc4SyxqRrQleZyDIHlc7bxnd


Downloading...
From: https://drive.google.com/uc?id=1s2TkjSpzNc4SyxqRrQleZyDIHlc7bxnd
To: /content/movies.csv
100% 112k/112k [00:00<00:00, 3.71MB/s]


In [ ]:
!gdown 1Ws-_s1fHZ9nHfGLVUQurbHDvStePlEJm

Downloading...
From: https://drive.google.com/uc?id=1Ws-_s1fHZ9nHfGLVUQurbHDvStePlEJm
To: /content/directors.csv
100% 65.4k/65.4k [00:00<00:00, 1.37MB/s]


#### Reading the dataset

Here we have two csv files 'movies.csv' and 'directors.csv' in the data.

Lets read both the csv files using pd.read_csv()

In [ ]:
movies = pd.read_csv('/content/movies.csv')
movies.head()

Unnamed: 0     id     budget  popularity     revenue  \
0           0  43597  237000000         150  2787965087   
1           1  43598  300000000         139   961000000   
2           2  43599  245000000         107   880674609   
3           3  43600  250000000         112  1084939099   
4           5  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
4         4767  2007   May   Tuesday

In [ ]:
directors = pd.read_csv('/content/directors.csv')


#### So what kind of questions can we ask from this dataset?
   - Since have `popularity`, can find, "Top 10 most popular
   movies"
   - Or using `vote_average`, I can find some highest rated movies also
   - Since, `year` is given, we can find number of movies released per year
   - And may be using both `budget` and `year`, I can find highest budget movies in a year

#### But can we ask more interesting/deeper questions?
   - Do you think we can find the most productive director?
   - Which director produces high budget films?
   - Highest and lowest rated movies for every month in a particular year

So, lets explore the dataset


Notice, that we also get a column **Unnamed: 0** which represents nothing but the index of a row.

Inorder to get rid of this column-

We can simply add one more argument `index_col=0`

The default value is index_col=None.

If we set index_col=0 we're explicitly stating to treat the first column as the index

In [ ]:
movies = pd.read_csv('movies.csv', index_col=0)
movies.head()

id     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
5  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
5                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
5         4767  2007   May   Tuesday


  - The movies df contains 1465 rows,11 columns


In [ ]:
movies.shape

(1465, 11)

In [ ]:
directors = pd.read_csv('directors.csv',index_col=0)
directors.head()

director_name    id gender
0      James Cameron  4762   Male
1     Gore Verbinski  4763   Male
2         Sam Mendes  4764   Male
3  Christopher Nolan  4765   Male
4     Andrew Stanton  4766   Male

In [ ]:
directors.shape

(2349, 3)

Now directors df contains:

- 2349 rows,3 columns

#### Summary

1. Movie dataset contains info about movies, release, popularity, ratings and the director ID
2. Director dataset contains detailed info about the director

Now, if we want to know the details about Director of a particular movie, we may have to join these datasets


### Merging of both Dataframe:

#### So we want to include directors df info into movies df. How can we do this ?

  - We can do it using `merge()`
  - Recall what is `merge()`?

    As you have seen in previous lecture:
    - The merge() method enables us to combine two dataframes together,
    - But it requires specifying the columns as a merge key.

#### Question: So on which column we should merge the dfs ?

- We will merge both the dataframes into single dataframe on the ID columns (representing unique director) in both the datasets
- If you look at the values of columns `director_id` of movies df and `id` of directors df
- You will observe that values of id in movies are referred or taken from directors df

But, first before merging, lets check if for all the movies in `movies` df, we have their corresponding director details present in the `directors` df or not.

For that, lets first check number of unique director values in our `movies` data




 Movie df contains 199 unique values of director id

 Lets check number of unique diretors in `directors` df

In [ ]:
directors['id'].nunique()

2349

- Movies Dataset: 1465 rows, but only 199 unique directors
- Directors Dataset: 2349 unique directors (= no of rows)


#### What can we infer from this?

- Looks like directors in `movies` is a subset of directors in `directors`
- But, still need to check if we have details for 199 directors present in `directors` df also

#### How to check whether all the values in director_id column of movies is present in id column of director ?

We can do it using `isin()` method of pandas

#### Cant we do this using Python `in`?

- We can, but this will work for one element at a time.
- We need to do this for all the values in the column
- The `isin()` method checks if the Dataframe column contains the specified value(s).



  

In [ ]:
movies['director_id'].isin(directors['id'])
#directors['id'].isin(movies['director_id'])

0       True
1       True
2       True
3       True
5       True
        ... 
4736    True
4743    True
4748    True
4749    True
4768    True
Name: director_id, Length: 1465, dtype: bool

- Notice that this is like a boolean "mask"
- It returns a df similar to the original df,
- where rows are checked and a boolean series is returned which is True
- wherever value of movies['director_id'] is present in directors['id'].

Lets see if there is any False here.





In [ ]:
np.all(movies['director_id'].isin(directors['id']))

np.True_

#### Lets finally merge the dataframes

#### Do we need to keep all the rows for movies?

Yes

#### Do we need to keep all the rows of directors?

No, only the ones for which we have a corresponding row in movies

###Question: So which `merge()` method do you think we should apply here ?

We can use  LEFT OUTER JOIN

#### Recall what will Left Outer Join do?

<img src="https://www.datasciencemadesimple.com/wp-content/uploads/2017/09/join-or-merge-in-python-pandas-1.png?ezimgfmt=ng:webp/ngcb1" />

It will include all the rows of df `movies` and only those from `directors` that match with values of `movies`['director_id']



Lets perform merge() using LEFT OUTER JOIN


In [ ]:
# if column name is not same
# `left_on`: Specifies the key of the 1st dataframe
# `right_on`: Specifies the key of the 2nd dataframe
data = movies.merge(directors, how='left', left_on='director_id',right_on='id')
data

id_x     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
4     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
1460  48363          0           3      321952   
1461  48370      27000          19     3151130   
1462  48375          0           7           0   
1463  48376          0           3           0   
1464  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      director_id  year month       day      director_name  id_y gender  
0            4762  2009   Dec  Thursday      James Cameron  4762   Male  
1            4763  2007   May  Saturday     Gore Verbinski  4763   Male  
2            4764  2015   Oct    Monday         Sam Mendes  4764   Male  
3            4765  2012   Jul    Monday  Christopher Nolan  4765   Male  
4            4767  2007   May   Tuesday          Sam Raimi  4767   Male  
...           ...   ...   ...       ...                ...   ...    ...  
1460         4809  1978   May    Monday    Martin Scorsese  4809   Male  
1461         5369  1994   Sep   Tuesday        Kevin Smith  5369   Male  
1462         5148  2009   Aug    Friday           Uwe Boll  5148   Male  
1463         5535  1990   Jul    Friday  Richard Linklater  5535   Male  
1464         5097  1992   Sep    Friday   Robert Rodriguez  5097    NaN  

[1465 rows x 14 columns]



### Question:  Do you think any column is redundant here and can be dropped?

  - `id_y` is redundant as it is same as `director_id`
  - But we dont require  `director_id` further

So we can simply drop these features

In [ ]:
data.drop(['director_id','id_y'],axis=1,inplace=True)
data.head()

id_x     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
4  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   
2                                   Spectre           6.3        4466  2015   
3                     The Dark Knight Rises           7.6        9106  2012   
4                              Spider-Man 3           5.9        3576  2007   

  month       day      director_name gender  
0   Dec  Thursday      James Cameron   Male  
1   May  Saturday     Gore Verbinski   Male  
2   Oct    Monday         Sam Mendes   Male  
3   Jul    Monday  Christopher Nolan   Male  
4   May   Tuesday          Sam Raimi   Male

## Feature Exploration


Lets explore all the features in the merged dataset



In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id_x           1465 non-null   int64  
 1   budget         1465 non-null   int64  
 2   popularity     1465 non-null   int64  
 3   revenue        1465 non-null   int64  
 4   title          1465 non-null   object 
 5   vote_average   1465 non-null   float64
 6   vote_count     1465 non-null   int64  
 7   year           1465 non-null   int64  
 8   month          1465 non-null   object 
 9   day            1465 non-null   object 
 10  director_name  1465 non-null   object 
 11  gender         1341 non-null   object 
dtypes: float64(1), int64(6), object(5)
memory usage: 137.5+ KB


- Looks like only `gender` column has missing values (will come later)

#### Lets describe these features to know more about their range of values

In [ ]:
data.describe()

id_x        budget   popularity       revenue  vote_average  \
count   1465.000000  1.465000e+03  1465.000000  1.465000e+03   1465.000000   
mean   45225.191126  4.802295e+07    30.855973  1.432539e+08      6.368191   
std     1189.096396  4.935541e+07    34.845214  2.064918e+08      0.818033   
min    43597.000000  0.000000e+00     0.000000  0.000000e+00      3.000000   
25%    44236.000000  1.400000e+07    11.000000  1.738013e+07      5.900000   
50%    45022.000000  3.300000e+07    23.000000  7.578164e+07      6.400000   
75%    45990.000000  6.600000e+07    41.000000  1.792469e+08      6.900000   
max    48395.000000  3.800000e+08   724.000000  2.787965e+09      8.300000   

         vote_count         year  
count   1465.000000  1465.000000  
mean    1146.396587  2002.615017  
std     1578.077438     8.680141  
min        1.000000  1976.000000  
25%      216.000000  1998.000000  
50%      571.000000  2004.000000  
75%     1387.000000  2009.000000  
max    13752.000000  2016.000000

In [ ]:
data.describe(include=object)

title month     day     director_name gender
count          1465  1465    1465              1465   1341
unique         1465    12       7               199      2
top     El Mariachi   Dec  Friday  Steven Spielberg   Male
freq              1   193     654                26   1309

- Generally budget and revenue for Hollywood movies is in million dollars
- But notice, the range of values in the `revenue` and `budget` seem to be very high


So it will be better to change the values into `million dollars USD`

#### How will you change the values of `revenue` and `budget` into `million dollars USD?


In [ ]:
data['revenue'] = (data['revenue']/1000000).round(2)
data


id_x     budget  popularity  revenue  \
0     43597  237000000         150  2787.97   
1     43598  300000000         139   961.00   
2     43599  245000000         107   880.67   
3     43600  250000000         112  1084.94   
4     43602  258000000         115   890.87   
...     ...        ...         ...      ...   
1460  48363          0           3     0.32   
1461  48370      27000          19     3.15   
1462  48375          0           7     0.00   
1463  48376          0           3     0.00   
1464  48395     220000          14     2.04   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      year month       day      director_name gender  
0     2009   Dec  Thursday      James Cameron   Male  
1     2007   May  Saturday     Gore Verbinski   Male  
2     2015   Oct    Monday         Sam Mendes   Male  
3     2012   Jul    Monday  Christopher Nolan   Male  
4     2007   May   Tuesday          Sam Raimi   Male  
...    ...   ...       ...                ...    ...  
1460  1978   May    Monday    Martin Scorsese   Male  
1461  1994   Sep   Tuesday        Kevin Smith   Male  
1462  2009   Aug    Friday           Uwe Boll   Male  
1463  1990   Jul    Friday  Richard Linklater   Male  
1464  1992   Sep    Friday   Robert Rodriguez    NaN  

[1465 rows x 12 columns]

In [ ]:
data['budget']=(data['budget']/1000000).round(2)
data.head()

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   
2  43599   245.0         107   880.67   
3  43600   250.0         112  1084.94   
4  43602   258.0         115   890.87   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   
2                                   Spectre           6.3        4466  2015   
3                     The Dark Knight Rises           7.6        9106  2012   
4                              Spider-Man 3           5.9        3576  2007   

  month       day      director_name gender  
0   Dec  Thursday      James Cameron   Male  
1   May  Saturday     Gore Verbinski   Male  
2   Oct    Monday         Sam Mendes   Male  
3   Jul    Monday  Christopher Nolan   Male  
4   May   Tuesday          Sam Raimi   Male

Now, finally our dataset is ready for analysis

Lets query the dataset to ask some interesting questions

## Fetching queries from dataframe

#### Lets say we are interested in fetching all highly rates movies

Say, I define any movie to be high rated when ratings > 7

#### How can we do this we get movies with ratings > 7?
-  Have you ever come across or seen SQL?
- For now, just know that **SQL is used to fetch data from databases**
- We can use basic **Pandas operations** in a similar way to **fetch desired data from loaded data**


### Masking

Lets first create a mask to filter such movies
- In SQL: SELECT * FROM movies WHERE vote_average > 7
- In pandas:

In [ ]:
data['vote_average'] > 7

0        True
1       False
2       False
3        True
4       False
        ...  
1460     True
1461     True
1462    False
1463    False
1464    False
Name: vote_average, Length: 1465, dtype: bool

But we still don't know the row values ... Only that which row satisfied the condtion

#### How do we get the row values then ?
  - By applying `movies.loc`
  - This is known as filtering

In [ ]:
data.loc[data['vote_average'] > 7]

id_x  budget  popularity  revenue  \
0     43597  237.00         150  2787.97   
3     43600  250.00         112  1084.94   
14    43616  250.00         120   956.02   
16    43619  250.00          94   958.40   
19    43622  200.00         100  1845.03   
...     ...     ...         ...      ...   
1456  48321    0.01          20     7.00   
1457  48323    0.00           5     0.00   
1458  48335    0.06          27     3.22   
1460  48363    0.00           3     0.32   
1461  48370    0.03          19     3.15   

                                          title  vote_average  vote_count  \
0                                        Avatar           7.2       11800   
3                         The Dark Knight Rises           7.6        9106   
14    The Hobbit: The Battle of the Five Armies           7.1        4760   
16          The Hobbit: The Desolation of Smaug           7.6        4524   
19                                      Titanic           7.5        7562   
...                                         ...           ...         ...   
1456                                 Eraserhead           7.5         485   
1457                                 The Mighty           7.1          51   
1458                                         Pi           7.1         586   
1460                             The Last Waltz           7.9          64   
1461                                     Clerks           7.4         755   

      year month        day      director_name gender  
0     2009   Dec   Thursday      James Cameron   Male  
3     2012   Jul     Monday  Christopher Nolan   Male  
14    2014   Dec  Wednesday      Peter Jackson   Male  
16    2013   Dec  Wednesday      Peter Jackson   Male  
19    1997   Nov    Tuesday      James Cameron   Male  
...    ...   ...        ...                ...    ...  
1456  1977   Mar   Saturday        David Lynch   Male  
1457  1998   Oct     Friday      Peter Chelsom   Male  
1458  1998   Jul     Friday   Darren Aronofsky   Male  
1460  1978   May     Monday    Martin Scorsese   Male  
1461  1994   Sep    Tuesday        Kevin Smith   Male  

[301 rows x 12 columns]

- Here, we will **only get those rows for which `data['vote_average'] > 7`**
- If we save the result back in `data`, **all other rows will be deleted**
- We might **still need the original data to work with**
- So it always a safe option to create a copy of your dataframe using **copy()** and perform any analysis using the copy

In [ ]:
df = data.copy(deep=True)
df

id_x  budget  popularity  revenue  \
0     43597  237.00         150  2787.97   
1     43598  300.00         139   961.00   
2     43599  245.00         107   880.67   
3     43600  250.00         112  1084.94   
4     43602  258.00         115   890.87   
...     ...     ...         ...      ...   
1460  48363    0.00           3     0.32   
1461  48370    0.03          19     3.15   
1462  48375    0.00           7     0.00   
1463  48376    0.00           3     0.00   
1464  48395    0.22          14     2.04   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      year month       day      director_name gender  
0     2009   Dec  Thursday      James Cameron   Male  
1     2007   May  Saturday     Gore Verbinski   Male  
2     2015   Oct    Monday         Sam Mendes   Male  
3     2012   Jul    Monday  Christopher Nolan   Male  
4     2007   May   Tuesday          Sam Raimi   Male  
...    ...   ...       ...                ...    ...  
1460  1978   May    Monday    Martin Scorsese   Male  
1461  1994   Sep   Tuesday        Kevin Smith   Male  
1462  2009   Aug    Friday           Uwe Boll   Male  
1463  1990   Jul    Friday  Richard Linklater   Male  
1464  1992   Sep    Friday   Robert Rodriguez    NaN  

[1465 rows x 12 columns]

In [ ]:
df[df['vote_average'] > 7]

id_x  budget  popularity  revenue  \
0     43597  237.00         150  2787.97   
3     43600  250.00         112  1084.94   
14    43616  250.00         120   956.02   
16    43619  250.00          94   958.40   
19    43622  200.00         100  1845.03   
...     ...     ...         ...      ...   
1456  48321    0.01          20     7.00   
1457  48323    0.00           5     0.00   
1458  48335    0.06          27     3.22   
1460  48363    0.00           3     0.32   
1461  48370    0.03          19     3.15   

                                          title  vote_average  vote_count  \
0                                        Avatar           7.2       11800   
3                         The Dark Knight Rises           7.6        9106   
14    The Hobbit: The Battle of the Five Armies           7.1        4760   
16          The Hobbit: The Desolation of Smaug           7.6        4524   
19                                      Titanic           7.5        7562   
...                                         ...           ...         ...   
1456                                 Eraserhead           7.5         485   
1457                                 The Mighty           7.1          51   
1458                                         Pi           7.1         586   
1460                             The Last Waltz           7.9          64   
1461                                     Clerks           7.4         755   

      year month        day      director_name gender  
0     2009   Dec   Thursday      James Cameron   Male  
3     2012   Jul     Monday  Christopher Nolan   Male  
14    2014   Dec  Wednesday      Peter Jackson   Male  
16    2013   Dec  Wednesday      Peter Jackson   Male  
19    1997   Nov    Tuesday      James Cameron   Male  
...    ...   ...        ...                ...    ...  
1456  1977   Mar   Saturday        David Lynch   Male  
1457  1998   Oct     Friday      Peter Chelsom   Male  
1458  1998   Jul     Friday   Darren Aronofsky   Male  
1460  1978   May     Monday    Martin Scorsese   Male  
1461  1994   Sep    Tuesday        Kevin Smith   Male  

[301 rows x 12 columns]

#### But this is not recommended. Why ?
  - It can create a confusion between implicit/explicit indexing used as discussed before
  - `loc` is also much faster

Now this is how we can start quering our data based on the conditions we want

#### We can also return only the subsets of columns

- Works **just like slicing**

In [ ]:
df.loc[df['vote_average'] > 7, ['title','vote_average']]
# These will be the only 2 columns printed out

title  vote_average
0                                        Avatar           7.2
3                         The Dark Knight Rises           7.6
14    The Hobbit: The Battle of the Five Armies           7.1
16          The Hobbit: The Desolation of Smaug           7.6
19                                      Titanic           7.5
...                                         ...           ...
1456                                 Eraserhead           7.5
1457                                 The Mighty           7.1
1458                                         Pi           7.1
1460                             The Last Waltz           7.9
1461                                     Clerks           7.4

[301 rows x 2 columns]

So far we saw only single condition for filtering

#### What if we have multiple conditions to filter rows?

#### What if we want to filter highly rated latest movies?

Notice that two conditions are involved here
1. Movies need to be highly rated i.e.. > 7
2. It should be recent, say released after 2014
- In SQL: SELECT * FROM movies WHERE vote_average > 7 AND year >= 2015
- We can **use AND operator b/w multiple conditions**

#### Recall how we apply mutliple conditions in numpy ?

- Use elementwise operator `&` or `|`
- Remember, **we cannot use `and` or `or` with dataframe as a  dataframe has multiple values**
- Also, recall, **for multiple conditions, we need to put each separate condition within parenthesis `()`**

In [ ]:
df.loc[(df['vote_average'] > 7) & (df['year'] >= 2015)].head()

id_x  budget  popularity  revenue                    title  \
30   43641   190.0         102  1506.25                Furious 7   
78   43724   150.0         434   378.86       Mad Max: Fury Road   
106  43773   135.0         100   532.95             The Revenant   
162  43867   108.0         167   630.16              The Martian   
312  44128    75.0          48   108.15  The Man from U.N.C.L.E.   

     vote_average  vote_count  year month        day  \
30            7.3        4176  2015   Apr  Wednesday   
78            7.2        9427  2015   May  Wednesday   
106           7.3        6396  2015   Dec     Friday   
162           7.6        7268  2015   Sep  Wednesday   
312           7.1        2265  2015   Aug   Thursday   

                   director_name gender  
30                     James Wan   Male  
78                 George Miller   Male  
106  Alejandro González Iñárritu   Male  
162                 Ridley Scott   Male  
312                  Guy Ritchie   Male

####Question: Get me all the movies which are alphabetically before movie title 'Avengers'

In [ ]:
df.loc[df['title'] < 'Avengers']
# String comparisons like this (>, <, ==) are also possible

id_x  budget  popularity  revenue                           title  \
0     43597   237.0         150  2787.97                          Avatar   
23    43629   200.0          78  1025.49             Alice in Wonderland   
40    43656   200.0          45   769.65                            2012   
41    43657   200.0          39   325.23               A Christmas Carol   
69    43709   155.0          39   167.30                       Alexander   
...     ...     ...         ...      ...                             ...   
1390  47491     3.0           7    20.97              A Room with a View   
1395  47575     3.0           3     0.00                        Amnesiac   
1405  47686     2.0          23    20.91                   Amores perros   
1432  47970     0.0           3     0.00              All the Real Girls   
1440  48155     0.0           0     0.00  Alleluia! The Devil's Carnival   

      vote_average  vote_count  year month        day  \
0              7.2       11800  2009   Dec   Thursday   
23             6.4        4645  2010   Mar  Wednesday   
40             5.6        4903  2009   Oct   Saturday   
41             6.6        1095  2009   Nov  Wednesday   
69             5.6         927  2004   Nov     Sunday   
...            ...         ...   ...   ...        ...   
1390           6.9         156  1985   Dec     Friday   
1395           4.1          52  2015   Aug     Friday   
1405           7.6         521  2000   Jun     Friday   
1432           5.9          30  2003   Aug     Friday   
1440           6.0           2  2016   Mar    Tuesday   

                    director_name gender  
0                   James Cameron   Male  
23                     Tim Burton   Male  
40                Roland Emmerich   Male  
41                Robert Zemeckis   Male  
69                   Oliver Stone   Male  
...                           ...    ...  
1390                  James Ivory   Male  
1395               Michael Polish   Male  
1405  Alejandro González Iñárritu   Male  
1432           David Gordon Green   Male  
1440          Darren Lynn Bousman   Male  

[125 rows x 12 columns]

### String methods in pandas
#### What kind of questions can be use string methods for?

Find rows which contains a particular string

Lets say you want to filter movies that contain pattern or substring 'The' in it

####How you can you filters row which has "The" in their movie titles?

- To apply a string method to a column, we will be using the `str` attribute of the Series object.


So in general, we will be using the following format:



     > Series.str.function()

- Series.str can be used to access the values of the series as strings and apply several methods to it.




- First we would need to access that series (or column), then add .str, and finally add the specific method we want to use.


Here coming back to our task:

- we can find the pattern 'The' in movies using str.contains()

- str.contains() function is used to test if pattern is contained within a string of a Series



In [ ]:
df['title'].str.contains('The')

0       False
1       False
2       False
3        True
4       False
        ...  
1460     True
1461    False
1462    False
1463    False
1464    False
Name: title, Length: 1465, dtype: bool

In [ ]:
df.loc[df['title'].str.contains('The')]

id_x  budget  popularity  revenue  \
3     43600  250.00         112  1084.94   
9     43610  255.00          49    89.29   
11    43612  225.00          53   419.65   
14    43616  250.00         120   956.02   
16    43619  250.00          94   958.40   
...     ...     ...         ...      ...   
1440  48155    0.00           0     0.00   
1443  48192    0.35          35    29.40   
1449  48244    0.25           6     0.06   
1457  48323    0.00           5     0.00   
1460  48363    0.00           3     0.32   

                                          title  vote_average  vote_count  \
3                         The Dark Knight Rises           7.6        9106   
9                               The Lone Ranger           5.9        2311   
11     The Chronicles of Narnia: Prince Caspian           6.3        1630   
14    The Hobbit: The Battle of the Five Armies           7.1        4760   
16          The Hobbit: The Desolation of Smaug           7.6        4524   
...                                         ...           ...         ...   
1440             Alleluia! The Devil's Carnival           6.0           2   
1443                              The Evil Dead           7.3         894   
1449                                The Canyons           4.1          75   
1457                                 The Mighty           7.1          51   
1460                             The Last Waltz           7.9          64   

      year month        day        director_name gender  
3     2012   Jul     Monday    Christopher Nolan   Male  
9     2013   Jul  Wednesday       Gore Verbinski   Male  
11    2008   May   Thursday       Andrew Adamson   Male  
14    2014   Dec  Wednesday        Peter Jackson   Male  
16    2013   Dec  Wednesday        Peter Jackson   Male  
...    ...   ...        ...                  ...    ...  
1440  2016   Mar    Tuesday  Darren Lynn Bousman   Male  
1443  1981   Oct   Thursday            Sam Raimi   Male  
1449  2013   Jul     Monday        Paul Schrader    NaN  
1457  1998   Oct     Friday        Peter Chelsom   Male  
1460  1978   May     Monday      Martin Scorsese   Male  

[327 rows x 12 columns]

In [ ]:
df.loc[df['title'].str.startswith('Batman')]

id_x  budget  popularity  revenue                               title  \
5    43606   250.0         155   873.26  Batman v Superman: Dawn of Justice   
74   43716   150.0         115   374.22                       Batman Begins   
128  43807   125.0          50   238.21                      Batman & Robin   
184  43896   100.0          48   336.53                      Batman Forever   
257  44025    80.0          59   280.00                      Batman Returns   
704  44956    35.0          44   411.35                              Batman   

     vote_average  vote_count  year month        day      director_name gender  
5             5.7        7004  2016   Mar  Wednesday        Zack Snyder   Male  
74            7.5        7359  2005   Jun     Friday  Christopher Nolan   Male  
128           4.2        1418  1997   Jun     Friday    Joel Schumacher   Male  
184           5.2        1498  1995   May  Wednesday    Joel Schumacher   Male  
257           6.6        1673  1992   Jun     Friday         Tim Burton   Male  
704           7.0        2096  1989   Jun     Friday         Tim Burton   Male

So, from whateve we have learnt so far, we can answer a couple of questions

### Question: How will you find Top 5 most popular movies?

- we can simply sort our data based on values of column 'popularity'

In [ ]:
df.sort_values(['popularity'] , ascending=False).head(5)
# df.sort_values(['popularity'],ascending=False).head(5)

id_x  budget  popularity  revenue  \
58   43692   165.0         724   675.12   
78   43724   150.0         434   378.86   
119  43796   140.0         271   655.01   
120  43797   125.0         206   752.10   
45   43662   185.0         187  1004.56   

                                                 title  vote_average  \
58                                        Interstellar           8.1   
78                                  Mad Max: Fury Road           7.2   
119  Pirates of the Caribbean: The Curse of the Bla...           7.5   
120              The Hunger Games: Mockingjay - Part 1           6.6   
45                                     The Dark Knight           8.2   

     vote_count  year month        day      director_name gender  
58        10867  2014   Nov  Wednesday  Christopher Nolan   Male  
78         9427  2015   May  Wednesday      George Miller   Male  
119        6985  2003   Jul  Wednesday     Gore Verbinski   Male  
120        5584  2014   Nov    Tuesday   Francis Lawrence   Male  
45        12002  2008   Jul  Wednesday  Christopher Nolan   Male

#### Okay, so who is your favorite Director?

My favourite director is **Christopher Nolan**

Lets say I am interested in getting names of all movies directed by my favourite director

#### Question: How will get list of movies directed by 'Christopher Nolan'?

In [ ]:
df.loc[df['director_name'] == 'Christopher Nolan',['title']]

title
58             Interstellar
45          The Dark Knight
59                Inception
74            Batman Begins
3     The Dark Knight Rises
641            The Prestige
1341                Memento
565                Insomnia

#### Quiz2:


1. How to get the  details of movies released in the month of 'Jan' or 'Nov' from the dataset ?

  a. df.loc[(df['month']='Jan') | (df['month']='Nov')]

  b. df.loc[(df['month']=='Jan') || (df['month']=='Nov')]

  c. df.loc[df['month']=='Jan'| df['month']=='Nov']

  d. df.loc[(df['month']=='Jan') | (df['month']=='Nov')]

  Ans: d

#### Quiz3:

2. How to filter those records where movies released in the year(2015,2016,2012) from the above dataset ?

 a. df['year'].isin([2015, 2016, 2012])

 b. df['year'].in([2015, 2016, 2012])

 c. df['year']==([2015, 2016, 2012])

Ans:a







In [ ]:
df.loc[(df['month']=='Jan') | (df['month']=='Nov')]

id_x  budget  popularity  revenue  \
58    43692   165.0         724   675.12   
120   43797   125.0         206   752.10   
167   43873   100.0         132   876.69   
695   44934    37.0         127   392.62   
63    43699   160.0         127   653.43   
...     ...     ...         ...      ...   
991   45670     0.0           1    37.00   
1438  48145     0.5           0     0.00   
1448  48233     0.0           0     0.00   
1164  46234     0.0           0     0.00   
1437  48135     0.0           0     0.00   

                                        title  vote_average  vote_count  year  \
58                               Interstellar           8.1       10867  2014   
120     The Hunger Games: Mockingjay - Part 1           6.6        5584  2014   
167   Harry Potter and the Chamber of Secrets           7.4        5815  2002   
695                                  Twilight           5.8        3561  2008   
63      The Hunger Games: Mockingjay - Part 2           6.6        3984  2015   
...                                       ...           ...         ...   ...   
991                         For Colored Girls           7.0          22  2010   
1438                  Everything Put Together           5.0           2  2001   
1448                             Chuck & Buck           5.7          16  2000   
1164                        Felicia's Journey           6.6          23  1999   
1437                       Finishing The Game           4.9           8  2007   

     month        day        director_name  gender  
58     Nov  Wednesday    Christopher Nolan    Male  
120    Nov    Tuesday     Francis Lawrence    Male  
167    Nov  Wednesday       Chris Columbus    Male  
695    Nov   Thursday  Catherine Hardwicke  Female  
63     Nov  Wednesday     Francis Lawrence    Male  
...    ...        ...                  ...     ...  
991    Nov     Friday          Tyler Perry    Male  
1438   Nov     Friday         Marc Forster    Male  
1448   Jan     Friday        Miguel Arteta    Male  
1164   Nov  Wednesday          Atom Egoyan    Male  
1437   Jan     Sunday           Justin Lin     NaN  

[177 rows x 12 columns]

In [ ]:
df['year'].isin([2015, 2016, 2012])

58      False
78       True
119     False
120     False
45      False
        ...  
1431    False
1437    False
1436    False
879     False
1378    False
Name: year, Length: 1465, dtype: bool

## Grouping

#### Question :Now lets say we want to know the number of movies released by a particular director

In [ ]:
df.loc[df['director_name'] == 'Christopher Nolan',['title']].count()

title    8
dtype: int64

#### What if we have to do this all possible directors?

I can probably do a value_counts() of the directors

In [ ]:
df["director_name"].value_counts()

director_name
Steven Spielberg     26
Martin Scorsese      19
Clint Eastwood       19
Woody Allen          18
Robert Rodriguez     16
                     ..
Nicole Holofcener     5
Peter Chelsom         5
Mike Leigh            5
Mira Nair             5
Michael Polish        5
Name: count, Length: 199, dtype: int64

In [ ]:
df.groupby('director_name')

In [ ]:
df.groupby('director_name')['budget'].max()

director_name
Adam McKay                     100.0
Adam Shankman                   80.0
Alejandro González Iñárritu    135.0
Alex Proyas                    140.0
Alexander Payne                 30.0
                               ...  
Wes Craven                      40.0
Wolfgang Petersen              175.0
Woody Allen                     30.0
Zack Snyder                    250.0
Zhang Yimou                     94.0
Name: budget, Length: 199, dtype: float64

In [ ]:
df.groupby('director_name').count()

id_x  budget  popularity  revenue  title  \
director_name                                                           
Adam McKay                      6       6           6        6      6   
Adam Shankman                   8       8           8        8      8   
Alejandro González Iñárritu     6       6           6        6      6   
Alex Proyas                     5       5           5        5      5   
Alexander Payne                 5       5           5        5      5   
...                           ...     ...         ...      ...    ...   
Wes Craven                     10      10          10       10     10   
Wolfgang Petersen               7       7           7        7      7   
Woody Allen                    18      18          18       18     18   
Zack Snyder                     7       7           7        7      7   
Zhang Yimou                     6       6           6        6      6   

                             vote_average  vote_count  year  month  day  \
director_name                                                             
Adam McKay                              6           6     6      6    6   
Adam Shankman                           8           8     8      8    8   
Alejandro González Iñárritu             6           6     6      6    6   
Alex Proyas                             5           5     5      5    5   
Alexander Payne                         5           5     5      5    5   
...                                   ...         ...   ...    ...  ...   
Wes Craven                             10          10    10     10   10   
Wolfgang Petersen                       7           7     7      7    7   
Woody Allen                            18          18    18     18   18   
Zack Snyder                             7           7     7      7    7   
Zhang Yimou                             6           6     6      6    6   

                             gender  
director_name                        
Adam McKay                        6  
Adam Shankman                     8  
Alejandro González Iñárritu       6  
Alex Proyas                       5  
Alexander Payne                   0  
...                             ...  
Wes Craven                       10  
Wolfgang Petersen                 7  
Woody Allen                      18  
Zack Snyder                       7  
Zhang Yimou                       6  

[199 rows x 11 columns]

In [ ]:
df.groupby('director_name')['title'].count()

director_name
Adam McKay                      6
Adam Shankman                   8
Alejandro González Iñárritu     6
Alex Proyas                     5
Alexander Payne                 5
                               ..
Wes Craven                     10
Wolfgang Petersen               7
Woody Allen                    18
Zack Snyder                     7
Zhang Yimou                     6
Name: title, Length: 199, dtype: int64

Now, lets say, you want to find who is the **most productive director**

#### Question: Which director according to you would be considered as most productive ?

Will you decide based on the number of movies released by a director?

Or will consider quality into consideration also?





Lets keep it simple for now, lets calculate who has directed maximum number of movies

In [ ]:
df.groupby(['director_name'])['title'].count().sort_values(ascending=False)

director_name
Steven Spielberg    26
Martin Scorsese     19
Clint Eastwood      19
Woody Allen         18
Robert Rodriguez    16
                    ..
Stephen Daldry       5
Tom Tykwer           5
Tim Hill             5
Uwe Boll             5
Wayne Wang           5
Name: title, Length: 199, dtype: int64

Looks like `Steven Spielberg` has directed maximum number of movies

#### But does it make him the most productive director?

Chances are, he might be active for more years than other directors

Lets calculate the number of active years of each director

#### How would you calculate active years for every director?

You would have to calculate both `min` and `max` of `year` and then subtract it.

We can calculate multiple aggregates in a single go using `aggregate()` function

In [ ]:
df_agg = df.groupby(['director_name'])[["title", "year"]].aggregate({"year":['min','max'], "title": "count"})
df_agg

year       title
                              min   max count
director_name                                
Adam McKay                   2004  2015     6
Adam Shankman                2001  2012     8
Alejandro González Iñárritu  2000  2015     6
Alex Proyas                  1994  2016     5
Alexander Payne              1999  2013     5
...                           ...   ...   ...
Wes Craven                   1984  2011    10
Wolfgang Petersen            1981  2006     7
Woody Allen                  1977  2013    18
Zack Snyder                  2004  2016     7
Zhang Yimou                  2002  2014     6

[199 rows x 3 columns]

In [ ]:
df_agg["year"]

min   max
director_name                          
Adam McKay                   2004  2015
Adam Shankman                2001  2012
Alejandro González Iñárritu  2000  2015
Alex Proyas                  1994  2016
Alexander Payne              1999  2013
...                           ...   ...
Wes Craven                   1984  2011
Wolfgang Petersen            1981  2006
Woody Allen                  1977  2013
Zack Snyder                  2004  2016
Zhang Yimou                  2002  2014

[199 rows x 2 columns]

#### How can we convert these back to only one level of columns?

Example: `year_min`, `year_max`, `title_count`

Notice that these are tuples, and we can just join them

In [ ]:
df_agg.columns = ['_'.join(col) for col in df_agg.columns]
df_agg

year_min  year_max  title_count
director_name                                               
Adam McKay                       2004      2015            6
Adam Shankman                    2001      2012            8
Alejandro González Iñárritu      2000      2015            6
Alex Proyas                      1994      2016            5
Alexander Payne                  1999      2013            5
...                               ...       ...          ...
Wes Craven                       1984      2011           10
Wolfgang Petersen                1981      2006            7
Woody Allen                      1977      2013           18
Zack Snyder                      2004      2016            7
Zhang Yimou                      2002      2014            6

[199 rows x 3 columns]

Columns look good, but we may want to turn back the row labels into a proper column as well

#### How can we convert row labels into columns?

- We can use `reset_index()`
- Each **row gets assigned a label number**

In [ ]:
df_agg.reset_index()

director_name  year_min  year_max  title_count
0                     Adam McKay      2004      2015            6
1                  Adam Shankman      2001      2012            8
2    Alejandro González Iñárritu      2000      2015            6
3                    Alex Proyas      1994      2016            5
4                Alexander Payne      1999      2013            5
..                           ...       ...       ...          ...
194                   Wes Craven      1984      2011           10
195            Wolfgang Petersen      1981      2006            7
196                  Woody Allen      1977      2013           18
197                  Zack Snyder      2004      2016            7
198                  Zhang Yimou      2002      2014            6

[199 rows x 4 columns]

Looks, PRETTY GOOD.

#### Now, using the new features, can we find the most productive director?

We need to know for how many years the director has been active.

Then, we can calculate rate of directing movies by `title_count`/`yrs_active`


#### So, lets calculate `yrs_active` first


In [ ]:
df_agg["yrs_active"] = df_agg["year_max"] - df_agg["year_min"]
df_agg

year_min  year_max  title_count  yrs_active
director_name                                                           
Adam McKay                       2004      2015            6          11
Adam Shankman                    2001      2012            8          11
Alejandro González Iñárritu      2000      2015            6          15
Alex Proyas                      1994      2016            5          22
Alexander Payne                  1999      2013            5          14
...                               ...       ...          ...         ...
Wes Craven                       1984      2011           10          27
Wolfgang Petersen                1981      2006            7          25
Woody Allen                      1977      2013           18          36
Zack Snyder                      2004      2016            7          12
Zhang Yimou                      2002      2014            6          12

[199 rows x 4 columns]

#### Now we can calculate the rate of directing movies and sort the values

In [ ]:
df_agg["movie_per_yr"] = df_agg["title_count"] / df_agg["yrs_active"]
df_agg.sort_values("movie_per_yr", ascending=False)

year_min  year_max  title_count  yrs_active  movie_per_yr
director_name                                                              
Tyler Perry           2006      2013            9           7      1.285714
Jason Friedberg       2006      2010            5           4      1.250000
Shawn Levy            2002      2014           11          12      0.916667
Robert Rodriguez      1992      2014           16          22      0.727273
Adam Shankman         2001      2012            8          11      0.727273
...                    ...       ...          ...         ...           ...
Lawrence Kasdan       1985      2012            5          27      0.185185
Luc Besson            1985      2014            5          29      0.172414
Michael Apted         1980      2010            5          30      0.166667
Robert Redford        1980      2010            5          30      0.166667
Sidney Lumet          1976      2006            5          30      0.166667

[199 rows x 5 columns]

So what is the question asking us to do?

#### We want to filter the rows based on some group property (director's max budget movie)

- The cases when we want to filter rows based on group properties (aggragates) is called  **Group Based |Filtering**
- We can group it by director and then use `groupby().filter` function
- **Rows from groups are filtered if they do not satisfy the boolean criterion** specified by func.

Now to filter values, we need to pass a filtering function to `filter()`

### Group based Transformation

Suppose, for every movie, we want to find out if it was an expensive movie for its director

#### How do we assess the budget of any movie wrt director?

May be we can subtract the average `budget` of a director from `budget` col, for each director

#### How can we do that ?
  - Group data acc to `director_name`
  - Calc its average `budget`
  - Subtract it from the data of that `director_name`
  - This process of changing data using group property is known as **Group based Transformation**


Just like `groupby().filter()`, we will use `grouby().transform()` function here

In [ ]:
def sub_avg(x):
  x -= x.mean()
  return x

df.groupby(['director_name'])["budget"].transform(sub_avg)

58      39.375000
78      80.755714
119    -18.142857
120      4.400000
45      59.375000
          ...    
1431    -1.950000
1437   -79.000000
1436    -4.200000
879    -46.000000
1378   -31.153846
Name: budget, Length: 1465, dtype: float64

Notice how some numbers for the movies which are of higher budget, the result will positive and for others, it will be negative

Lets make the problem a little more interesting and challenging

What if we want to subtract 1 col's mean from the other col mean?


#### What if we want to do the transformation of a column using some column's agrregate

Lets say, we want to filter the movies whose budget was even higher than the average revenue of the director from his other movies

#### How do we filter movies whose making were big risk to directors?

May be we can subtract the average `revenue` of a director from `budget` col, for each director

But we can use `transform` here as it expects only one column

We can use `apply()` function here

#### How can we do it ?
  - We again need to group data acc to `director_name`
  - Subtracting mean of `budget` from `revenue`
  - To do so we would need to apply a custom function
  - We can do so using the apply() method

Notice how some numbers for the movies which are of higher budget, the result will positive and for others, it will be negative

Lets make the problem a little more interesting and challenging

What if we want to subtract 1 col's mean from the other col mean?


#### What if we want to do the transformation of a column using some column's agrregate

Lets say, we want to filter the movies whose budget was even higher than the average revenue of the director from his other movies

#### How do we filter movies whose making were big risk to directors?

May be we can subtract the average `revenue` of a director from `budget` col, for each director

But we can use `transform` here as it expects only one column

We can use `apply()` function here

#### How can we do it ?
  - We again need to group data acc to `director_name`
  - Subtracting mean of `budget` from `revenue`
  - To do so we would need to apply a custom function
  - We can do so using the apply() method

In [ ]:
def func(x):
  x["risky"] = x["budget"] - x["revenue"].mean() >= 0
  return x
df_risky = df.groupby("director_name").apply(func)
df_risky

/tmp/ipython-input-3139350867.py:4: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  df_risky = df.groupby("director_name").apply(func)


id_x  budget  popularity  revenue  \
director_name                                            
Adam McKay    839   45301    28.0          57   133.35   
              505   44503    50.0          38   173.65   
              916   45443    26.0          29    90.57   
              176   43882   100.0          24   170.43   
              366   44236    65.0          22   128.11   
...                   ...     ...         ...      ...   
Zhang Yimou   1217  46460     0.0          21    92.86   
              192   43914    94.0          12    95.31   
              590   44692     0.0           9     0.00   
              1389  47489     0.0           6     0.00   
              1223  46493     0.0           1     0.00   

                                                    title  vote_average  \
director_name                                                             
Adam McKay    839                           The Big Short           7.3   
              505       Anchorman 2: The Legend Continues           6.0   
              916   Anchorman: The Legend of Ron Burgundy           6.7   
              176                          The Other Guys           6.1   
              366                           Step Brothers           6.5   
...                                                   ...           ...   
Zhang Yimou   1217                House of Flying Daggers           7.1   
              192                      The Flowers of War           7.1   
              590              Curse of the Golden Flower           6.6   
              1389                            Coming Home           6.9   
              1223       A Woman, a Gun and a Noodle Shop           4.8   

                    vote_count  year month        day director_name gender  \
director_name                                                                
Adam McKay    839         2607  2015   Dec     Friday    Adam McKay   Male   
              505          923  2013   Dec  Wednesday    Adam McKay   Male   
              916         1493  2004   Jul     Friday    Adam McKay   Male   
              176         1383  2010   Aug     Friday    Adam McKay   Male   
              366         1062  2008   Jul     Friday    Adam McKay   Male   
...                        ...   ...   ...        ...           ...    ...   
Zhang Yimou   1217         439  2004   May  Wednesday   Zhang Yimou   Male   
              192          187  2011   Dec   Thursday   Zhang Yimou   Male   
              590          203  2006   Dec   Thursday   Zhang Yimou   Male   
              1389          49  2014   May     Friday   Zhang Yimou   Male   
              1223          13  2009   Dec     Friday   Zhang Yimou   Male   

                    risky  
director_name              
Adam McKay    839   False  
              505   False  
              916   False  
              176   False  
              366   False  
...                   ...  
Zhang Yimou   1217  False  
              192    True  
              590   False  
              1389  False  
              1223  False  

[1465 rows x 13 columns]

In [ ]:
df_risky.loc[df_risky["risky"]]

id_x  budget  popularity  revenue  \
director_name                                                 
Andrzej Bartkowiak 349   44192    60.0          29    55.99   
Atom Egoyan        1194  46370    15.0          26     8.46   
                   1347  47224     5.0           7     3.26   
                   946   45538    25.0           4     0.00   
Brett Ratner       24    43630   210.0           3   459.36   
...                        ...     ...         ...      ...   
Uwe Boll           1058  45834    20.0           9    10.44   
                   944   45536    25.0           7     2.41   
                   1383  47453     3.5           4     0.00   
Wayne Wang         468   44419    55.0          19   154.91   
Zhang Yimou        192   43914    94.0          12    95.31   

                                               title  vote_average  \
director_name                                                        
Andrzej Bartkowiak 349                          Doom           5.0   
Atom Egoyan        1194                        Chloe           5.9   
                   1347          The Sweet Hereafter           6.8   
                   946          Where the Truth Lies           5.9   
Brett Ratner       24          X-Men: The Last Stand           6.3   
...                                              ...           ...   
Uwe Boll           1058            Alone in the Dark           3.1   
                   944                    BloodRayne           3.5   
                   1383  In the Name of the King III           3.3   
Wayne Wang         468             Maid in Manhattan           5.6   
Zhang Yimou        192            The Flowers of War           7.1   

                         vote_count  year month        day  \
director_name                                                
Andrzej Bartkowiak 349          609  2005   Oct   Thursday   
Atom Egoyan        1194         498  2009   Mar  Wednesday   
                   1347         103  1997   May  Wednesday   
                   946           66  2005   Oct     Friday   
Brett Ratner       24          3525  2006   May  Wednesday   
...                             ...   ...   ...        ...   
Uwe Boll           1058         173  2005   Jan     Friday   
                   944          118  2005   Oct   Saturday   
                   1383          19  2013   Dec     Friday   
Wayne Wang         468          485  2002   Dec     Friday   
Zhang Yimou        192          187  2011   Dec   Thursday   

                              director_name gender  risky  
director_name                                              
Andrzej Bartkowiak 349   Andrzej Bartkowiak   Male   True  
Atom Egoyan        1194         Atom Egoyan   Male   True  
                   1347         Atom Egoyan   Male   True  
                   946          Atom Egoyan   Male   True  
Brett Ratner       24          Brett Ratner   Male   True  
...                                     ...    ...    ...  
Uwe Boll           1058            Uwe Boll   Male   True  
                   944             Uwe Boll   Male   True  
                   1383            Uwe Boll   Male   True  
Wayne Wang         468           Wayne Wang    NaN   True  
Zhang Yimou        192          Zhang Yimou   Male   True  

[131 rows x 13 columns]

In [ ]:
df_risky.loc[df_risky["risky"]]

id_x  budget  popularity  revenue  \
director_name                                                 
Andrzej Bartkowiak 349   44192    60.0          29    55.99   
Atom Egoyan        1194  46370    15.0          26     8.46   
                   1347  47224     5.0           7     3.26   
                   946   45538    25.0           4     0.00   
Brett Ratner       24    43630   210.0           3   459.36   
...                        ...     ...         ...      ...   
Uwe Boll           1058  45834    20.0           9    10.44   
                   944   45536    25.0           7     2.41   
                   1383  47453     3.5           4     0.00   
Wayne Wang         468   44419    55.0          19   154.91   
Zhang Yimou        192   43914    94.0          12    95.31   

                                               title  vote_average  \
director_name                                                        
Andrzej Bartkowiak 349                          Doom           5.0   
Atom Egoyan        1194                        Chloe           5.9   
                   1347          The Sweet Hereafter           6.8   
                   946          Where the Truth Lies           5.9   
Brett Ratner       24          X-Men: The Last Stand           6.3   
...                                              ...           ...   
Uwe Boll           1058            Alone in the Dark           3.1   
                   944                    BloodRayne           3.5   
                   1383  In the Name of the King III           3.3   
Wayne Wang         468             Maid in Manhattan           5.6   
Zhang Yimou        192            The Flowers of War           7.1   

                         vote_count  year month        day  \
director_name                                                
Andrzej Bartkowiak 349          609  2005   Oct   Thursday   
Atom Egoyan        1194         498  2009   Mar  Wednesday   
                   1347         103  1997   May  Wednesday   
                   946           66  2005   Oct     Friday   
Brett Ratner       24          3525  2006   May  Wednesday   
...                             ...   ...   ...        ...   
Uwe Boll           1058         173  2005   Jan     Friday   
                   944          118  2005   Oct   Saturday   
                   1383          19  2013   Dec     Friday   
Wayne Wang         468          485  2002   Dec     Friday   
Zhang Yimou        192          187  2011   Dec   Thursday   

                              director_name gender  risky  
director_name                                              
Andrzej Bartkowiak 349   Andrzej Bartkowiak   Male   True  
Atom Egoyan        1194         Atom Egoyan   Male   True  
                   1347         Atom Egoyan   Male   True  
                   946          Atom Egoyan   Male   True  
Brett Ratner       24          Brett Ratner   Male   True  
...                                     ...    ...    ...  
Uwe Boll           1058            Uwe Boll   Male   True  
                   944             Uwe Boll   Male   True  
                   1383            Uwe Boll   Male   True  
Wayne Wang         468           Wayne Wang    NaN   True  
Zhang Yimou        192          Zhang Yimou   Male   True  

[131 rows x 13 columns]

Note: `apply()` can be applied on any dataframe along any particular axis

#### What does this mean ?
  - The custom func will be apllied on each row if axis = 0 and on each col if axis = 1
  - By default axis = 0
  - Lets create a new dataframe to understand this

In [ ]:
df[['revenue', 'budget']].apply(np.sum, axis = 0)

revenue    209867.04
budget      70353.62
dtype: float64

For each year , how you will get the average, lowest, and highest value of ratings of movies?

In [ ]:
df.groupby('year').agg({'vote_average': ['mean', 'min', 'max']})

vote_average          
             mean  min  max
year                       
1976     7.633333  7.2  8.0
1977     6.900000  6.1  7.8
1978     7.020000  5.9  7.9
1979     7.025000  5.6  8.0
1980     7.187500  6.4  7.9
1981     6.927273  6.0  7.9
1982     6.810000  5.5  7.9
1983     6.666667  4.5  8.0
1984     6.760000  6.1  7.3
1985     6.885714  5.9  8.0
1986     6.809091  5.8  7.8
1987     6.991667  6.1  7.6
1988     6.692308  5.7  7.6
1989     6.750000  5.5  7.6
1990     6.661111  5.8  8.2
1991     7.100000  6.6  8.1
1992     6.640909  5.7  8.0
1993     7.144444  6.1  8.3
1994     6.596774  5.0  8.3
1995     6.509091  4.5  8.1
1996     6.384783  5.4  7.8
1997     6.552174  4.1  7.9
1998     6.442222  4.9  7.9
1999     6.384375  5.1  8.3
2000     6.347458  4.4  8.1
2001     6.217742  3.7  8.0
2002     6.359420  3.4  8.0
2003     6.115517  4.4  8.1
2004     6.427869  4.7  7.7
2005     6.229577  3.1  7.7
2006     6.212329  3.6  8.0
2007     6.307407  3.2  7.9
2008     6.268657  3.0  8.2
2009     6.190909  3.9  7.9
2010     6.223810  4.2  8.1
2011     6.098507  4.1  7.7
2012     6.216327  4.9  7.8
2013     6.315789  3.3  7.9
2014     6.313208  4.1  8.1
2015     6.141463  3.9  7.7
2016     5.915385  4.5  7.1